Require passphrase to execute query

fractal161 opened the pull request:
Putting this up for review because I feel its current state is... not great.

The PR implements the really simple security protocol of asking for a passphrase in the header and then yelling at the client if they don't provide it. However, the way it's done now doesn't really lend itself to a refactor; the [helix docs](https://graphql-helix.vercel.app/docs/recipes/authentication-and-authorization) seem to recommend handling this all within the `contextFactory`, while I did stuff in the main fastify handler. However, I think doing it this way would mean explicitly handling perms for each API request, which is effort. Ideal situation would be throwing an error from `contextFactory` and having helix magically convert that into what we want, but I haven't really looked into this yet.

Anyways, would like feedback, not really sure how hard to go wrt all of this (but more than happy to put in the time if needed, don't have much else on my plate atm). Cheers.

fractal161 committed:
check for passphrase header before executing query

Figured it was sensible to keep the phrase as an environment variable.
This implementation just checks if the proper header is set. If no
header is provided or the value is incorrect, an appropriate graphql
error is sent.

When we move to a more robust authentication system, I think we will
want to move this logic into the contextFactory, which is recommended by
https://graphql-helix.vercel.app/docs/recipes/authentication-and-authorization.
However, right now I think this requires manually handling permissions
for every single query, so I think this should be saved for post-MVP.

fractal161 committed:
make tests send passphrase header

Did a blind find/replace across all of the files, this probably doesn't
break anything. Definitely highlights the need for a sensible refactor.

professor-l reviewed (approved):
I'm not really worried about whether we're implementing this "properly" (i.e. within `contextFactory`) because checking a hard-coded value is already a _really_ improper approach to authn/authz, and we'll replace this wholesale into a more robust approach in the future. This looks good for now, thank you again!

fractal161 committed:
fix dev repl instructions

fractal161 committed:
rename version to versionId in EloSnapshot

The EloSnapshot present in prisma stores the versionId, but the gql
schema asked specifically for `version`, which was the human-readable
form. Eventually, this should be refactored into a resolver, which can
expose all desired EloVersion details.

fractal161 committed:
allow passphrase disabling for development

motivated because graphiql stopped loading our schema because the
request it sent didn't include the passphrase lmao. having either an
empty string for PASSPHRASE or simply not including the field will
disable the check.

i think this motivates the creation of a .env.development file, so dev
and prod builds can handle this differently, but it should be fine for
now.
